File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/Storage.java
package com.example.alexeyglushkov.cachemanager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

/**
 * Created by alexeyglushkov on 26.09.15.
 */
// TODO: add exception throwing, create safe interface like SafeStorage with optGetValue/
// TODO: add iterator support
public interface Storage {
    void put(@NonNull String key, @NonNull Object value, @Nullable StorageMetadata metadata) throws Exception;

    @Nullable Object getValue(@NonNull String key) throws Exception;

    @NonNull StorageMetadata createMetadata();

    @Nullable StorageMetadata getMetadata(@NonNull String key) throws Exception;
    void remove(@NonNull String key) throws Exception;

    @Nullable StorageEntry getEntry(@NonNull String key) throws Exception;
    List<StorageEntry> getEntries() throws Exception;
    void removeAll() throws Exception;
}

File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/Storage.kt
package com.example.alexeyglushkov.cachemanager

/**
 * Created by alexeyglushkov on 26.09.15.
 */
// TODO: add exception throwing, create safe interface like SafeStorage with optGetValue/
// TODO: add iterator support
interface Storage {
    @Throws(Exception::class) fun put(key: String, value: Any, metadata: StorageMetadata?)

    @Throws(Exception::class) fun getValue(key: String): Any?

    fun createMetadata(): StorageMetadata

    @Throws(Exception::class) fun getMetadata(key: String): StorageMetadata?
    @Throws(Exception::class) fun remove(key: String)

    @Throws(Exception::class) fun getEntry(key: String): StorageEntry?
    @Throws(Exception::class) fun getEntries(): List<StorageEntry>

    @Throws(Exception::class) fun removeAll()
}